sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -468,14 +460,16 @@ object PartitioningUtils {
   }
 
   /**
-   * Given a collection of [[Literal]]s, resolves possible type conflicts by up-casting "lower"
-   * types.
+   * Given a collection of [[Literal]]s, resolves possible type conflicts by
+   * [[TypeCoercion.findWiderCommonType]]. See [[TypeCoercion.findWiderTypeForTwo]].
    */
   private def resolveTypeConflicts(literals: Seq[Literal], timeZone: TimeZone): Seq[Literal] = {
-    val desiredType = {
-      val topType = literals.map(_.dataType).maxBy(upCastingOrder.indexOf(_))
```

Comment:
I think it's a bug, the previous code didn't consider the case when input literal types are outside of the `upCastingOrder`, and just pick the first type as the final type.

However I'm not sure what's the expected behavior. We need to figure out what's the possible data types for partition columns, and how to merge them.

Reply:
Let me try to describe what this PR explicitly changes soon.

Reply:
I think we will have the input types for this `resolveTypeConflicts`:

```scala
Seq(
    NullType, IntegerType, LongType, DoubleType,
    *DecimalType(...), DateType, TimestampType, StringType)
```

*`DecimalType` only when it's bigger than `LongType`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L384-L393



Because:

this particular `resolveTypeConflicts` seems being only called through:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L142

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L337

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L474

In the first call, I am seeing `pathsWithPartitionValues` is constructed by `partitionValues`, which is the output from `parsePartition`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L108

which parses the input by `parsePartitionColumn`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L209

which calls this `inferPartitionColumnValue`:


https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L384-L428

Reply:
So, for the types:

```scala
Seq(
    NullType, IntegerType, LongType, DoubleType,
    *DecimalType(...), DateType, TimestampType, StringType)
```


I produced a chart as below by this codes:

```scala
test("Print out chart") {
  val supportedTypes: Seq[DataType] = Seq(
    NullType, IntegerType, LongType, DoubleType,
    DecimalType(38, 0), DateType, TimestampType, StringType)

  val combinations = for {
    t1 <- supportedTypes
    t2 <- supportedTypes
  } yield Seq(t1,t2)

  // Old type conflict resolution:
  val upCastingOrder: Seq[DataType] =
    Seq(NullType, IntegerType, LongType, FloatType, DoubleType, StringType)

  def oldResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
    val topType = dataTypes.maxBy(upCastingOrder.indexOf(_))
    if (topType == NullType) StringType else topType
  }

  // New type conflict resolution:
  def newResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
    TypeCoercion.findWiderCommonType(dataTypes) match {
      case Some(NullType) => StringType
      case Some(dt: DataType) => dt
      case _ => StringType
    }
  }

  println("|Input types|Old output type|New output type|")
  println("|-----------|---------------|---------------|")
  combinations.foreach { pair =>
    val oldType = oldResolveTypeConflicts(pair)
    val newType = newResolveTypeConflicts(pair)

    if (oldType != newType) {
      println(s"|[`${pair(0)}`, `${pair(1)}`]|`$oldType`|`$newType`|")
    }
  }
}
```

So, looks this PR makes the changes in type resolution as below:

|Input types|Old output type|New output type|
|-----------|---------------|---------------|
|[`NullType`, `DecimalType(38,0)`]|`StringType`|`DecimalType(38,0)`|
|[`NullType`, `DateType`]|`StringType`|`DateType`|
|[`NullType`, `TimestampType`]|`StringType`|`TimestampType`|
|[`IntegerType`, `DecimalType(38,0)`]|`IntegerType`|`DecimalType(38,0)`|
|[`IntegerType`, `DateType`]|`IntegerType`|`StringType`|
|[`IntegerType`, `TimestampType`]|`IntegerType`|`StringType`|
|[`LongType`, `DecimalType(38,0)`]|`LongType`|`DecimalType(38,0)`|
|[`LongType`, `DateType`]|`LongType`|`StringType`|
|[`LongType`, `TimestampType`]|`LongType`|`StringType`|
|[`DoubleType`, `DateType`]|`DoubleType`|`StringType`|
|[`DoubleType`, `TimestampType`]|`DoubleType`|`StringType`|
|[`DecimalType(38,0)`, `NullType`]|`StringType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `IntegerType`]|`IntegerType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `LongType`]|`LongType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `DateType`]|`DecimalType(38,0)`|`StringType`|
|[`DecimalType(38,0)`, `TimestampType`]|`DecimalType(38,0)`|`StringType`|
|[`DateType`, `NullType`]|`StringType`|`DateType`|
|[`DateType`, `IntegerType`]|`IntegerType`|`StringType`|
|[`DateType`, `LongType`]|`LongType`|`StringType`|
|[`DateType`, `DoubleType`]|`DoubleType`|`StringType`|
|[`DateType`, `DecimalType(38,0)`]|`DateType`|`StringType`|
|[`DateType`, `TimestampType`]|`DateType`|`TimestampType`|
|[`TimestampType`, `NullType`]|`StringType`|`TimestampType`|
|[`TimestampType`, `IntegerType`]|`IntegerType`|`StringType`|
|[`TimestampType`, `LongType`]|`LongType`|`StringType`|
|[`TimestampType`, `DoubleType`]|`DoubleType`|`StringType`|
|[`TimestampType`, `DecimalType(38,0)`]|`TimestampType`|`StringType`|

Reply:
I think the new behavior is much better. It seems like the previous behavior is just wrong, but it's very rare to see different data types in partition columns, and that's why no users open tickets for it yet.

cc @gatorsmile

Reply:
Since this changes the schema, we can't change it without a conf.

Reply:
We also need to document the behavior change in the doc.

Reply:
Do we really wanna document the previous wrong behavior in the doc? An example is merging `TimestampType` and `DateType`, the result is non-deterministic, depends on which partition path gets parsed first. How do we document that?

Reply:
BTW this is only used in partition discovery, I can't think of a problematic case for it. The only thing it can break is, users give Spark a data dir to do partition discovering, and users make an assumption on the inferred partition schema, but then we can argue that why users ask Spark to do partition discovery at the beginning?

Reply:
Both the values and the schemas could be changed. The external applications might be broken if the schema is different. 

The new behaviors are consistent with what we does for the other type coercion cases. However, implicit type casting and partition discovery are unstable. The other mature systems have clear/stable rules about it. Below is an example. https://docs.oracle.com/cloud/latest/db112/SQLRF/sql_elements002.htm#g195937

If each release introduces new behaviors, it becomes hard to use by the end users who have such expectation. Thus, my suggestion is to first stabilize our type coercion rules before addressing this. 

https://github.com/apache/spark/pull/18853 is the first PR attempt in this direction.

Reply:
Partitioned columns are different from normal type coercion cases, they are literally all string type, and we are just trying to find a most reasonable type of them.

The previous behavior was there since the very beginning, which I think didn't go through a decent discussion. This is the first time we seriously design the type merging logic for partition discovery. I think it doesn't need to be blocked by the type coercion stabilization work, as they can diverge.

@HyukjinKwon can you send the proposal to dev list? I think we need more feedback, e.g. people may want more strict rules and have more cases to fallback to string.

Reply:
Let me send the proposal to dev soon tonight (KST)